engine/src/main/java/io/zeebe/engine/processing/bpmn/BpmnStreamProcessor.java

```diff
@@ -127,7 +127,23 @@ public void processRecord(
           ProcessInstanceIntent.ELEMENT_ACTIVATING);
     }
 
-    // process the event
+    // migrated processors expect their state to be set via event appliers, and non migrated
+    // processors in the processor. this means that ELEMENT_COMPLETING of a migrated processor will
+    // set the element state to ELEMENT_COMPLETING on replay, but ELEMENT_COMPLETING of a non
+    // migrated will expect its state to ALREADY be ELEMENT_COMPLETING. To avoid too much
+    // complexity, when reprocessing events for non migrated processors, just set their state as
+    // expected; this could in the future be a source of error, but as its only temporary until
+    // all processors are migrated, it's OK for now
+    // TODO(npepinpe): remove as part of https://github.com/camunda-cloud/zeebe/issues/6202
+    if (!MigratedStreamProcessors.isMigrated(context.getBpmnElementType())
+        && (context.getIntent() == ProcessInstanceIntent.ELEMENT_COMPLETING
+            || context.getIntent() == ProcessInstanceIntent.ELEMENT_TERMINATING
+            || context.getIntent() == ProcessInstanceIntent.ELEMENT_TERMINATED)) {
+      elementInstanceState.updateInstance(
+          context.getElementInstanceKey(), e -> e.setState(context.getIntent()));
+    }
```

Comment:
This generally mimics the behavior we expect in the future, i.e. the event appliers will update the state of the element instance. However I'm worried this could cause issues/interaction problems with non migrated processors...

The alternatives I checked were imo worse:

1. Moving the child-completes-the-parent logic into the appliers (works but means after migration we need to move it back again to processing, so unnecessary work and adds logic to appliers)
2. Adding more information to the record itself (i.e. encode in the record that we terminated the parent, so we need to change their state accordingly when applying the event). Slightly better than 1, but still meh and needs to temporarily modify the record only to revert it after

Hopefully you have a better idea :smile: